File: 2022/typescript/16.ts
import {readFileSync} from 'fs'

type ValveID = string
type Valve = {id: ValveID; rate: number; tunnels: ValveID[]}
type ValveMap = Record<ValveID, Valve>

class Graph {
  private readonly valves: ValveMap
  private readonly cache: Record<string, number>

  constructor(valves: ValveMap) {
    this.valves = valves
    this.cache = {}
  }

  findMaxPressure(valve: Valve, minutes: number, opened: Set<ValveID>, others: 0 | 1): number {
    if (minutes === 0) {
      if (others) {
        return this.findMaxPressure(this.valves['AA'], 26, opened, 0)
      } else {
        return 0
      }
    }

    const key = this.computeCacheKey(valve, minutes, opened, others)

    if (this.cache[key] >= 0) {
      return this.cache[key]
    }

    let maxPressure = 0

    if (valve.rate && !opened.has(valve.id)) {
      maxPressure = Math.max(
        maxPressure,
        (minutes - 1) * valve.rate + this.findMaxPressure(valve, minutes - 1, new Set([...opened, valve.id]), others)
      )
    }

    for (const tunnel of valve.tunnels) {
      maxPressure = Math.max(maxPressure, this.findMaxPressure(this.valves[tunnel], minutes - 1, opened, others))
    }

    this.cache[key] = maxPressure

    return maxPressure
  }

  private computeCacheKey(valve: Valve, minutes: number, opened: Set<ValveID>, others: 0 | 1): string {
    return `${valve.id}${minutes}${[...opened].sort().join('')}${others}`
  }
}

const valves: ValveMap = readFileSync('./2022/16.input', 'utf8')
  .trim()
  .split(/\n/)
  .reduce((valves: ValveMap, line: string) => {
    const match = /Valve ([A-Z]{2}) has flow rate=(\d+); tunnels? leads? to valves? ([A-Z, ]+)/.exec(line)
    const id = match![1]
    const rate = Number(match![2])
    const tunnels = match![3].split(', ')
    valves[id] = {id, rate, tunnels}
    return valves
  }, {})

const graph = new Graph(valves)

const first = graph.findMaxPressure(valves['AA'], 30, new Set(), 0)
console.log(first)

const second = graph.findMaxPressure(valves['AA'], 26, new Set(), 1)
console.log(second)

File: 2022/python/16.py
import re
from collections import namedtuple
from functools import cache

Valve = namedtuple("Valve", "id, rate, tunnels")


@cache
def find_max_pressure(valve, minutes, opened, others):
    if minutes == 0:
        if others:
            return find_max_pressure(valves["AA"], 26, opened, 0)
        else:
            return 0

    max_pressure = 0

    if valve.rate and valve.id not in opened:
        max_pressure = max(
            max_pressure,
            (minutes - 1) * valve.rate + find_max_pressure(valve, minutes - 1, frozenset((*opened, valve.id)), others),
        )

    for tunnel in valve.tunnels:
        max_pressure = max(max_pressure, find_max_pressure(valves[tunnel], minutes - 1, opened, others))

    return max_pressure


valves = dict()

with open("./2022/16.input") as file:
    pattern = re.compile(r"Valve ([A-Z]{2}) has flow rate=(\d+); tunnels? leads? to valves? ([A-Z, ]+)")

    for line in file.read().splitlines():
        id, rate, tunnels = pattern.findall(line)[0]
        valves[id] = Valve(id, int(rate), tuple(tunnels.split(", ")))


first = find_max_pressure(valves["AA"], 30, (), 0)
print(first)

second = find_max_pressure(valves["AA"], 26, (), 1)
print(second)
